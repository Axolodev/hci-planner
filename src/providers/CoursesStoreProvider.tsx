import {
  CoursesStoreType,
  createCoursesStore,
  initCounterStore,
} from "@/stores/coursesStore";
import { createContext, ReactNode, useContext, useRef } from "react";
import { useStore } from "zustand";

// Define provides for React
export type CoursesStoreApi = ReturnType<typeof createCoursesStore>;

export const CoursesStoreContext = createContext<CoursesStoreApi | undefined>(
  undefined
);

export interface CoursesStoreProviderProps {
  children: ReactNode;
}

export const CoursesStoreProvider = ({
  children,
}: CoursesStoreProviderProps) => {
  const storeRef = useRef<CoursesStoreApi | null>(null);
  if (storeRef.current === null) {
    storeRef.current = createCoursesStore(initCounterStore());
  }

  return (
    <CoursesStoreContext.Provider value={storeRef.current}>
      {children}
    </CoursesStoreContext.Provider>
  );
};

// Define a hook to access the store
export const useCoursesStore = <T,>(
  selector: (store: CoursesStoreType) => T
): T => {
  const coursesStoreContext = useContext(CoursesStoreContext);

  if (!coursesStoreContext) {
    throw new Error(`useCoursesStore must be used within CoursesStoreProvider`);
  }

  return useStore(coursesStoreContext, selector);
};
