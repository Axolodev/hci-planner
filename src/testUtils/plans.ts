import { Plan } from "@/types";

export const basicPlan = {
  sections: [
    {
      title: "Section title",
      options: [
        {
          name: "Option name",
          modules: [
            {
              name: "Module name",
            },
          ],
        },
      ],
      description: "",
    },
  ],
};

export const planWithMultipleSectionOptions: Plan = {
  sections: [
    {
      title: "Section title",
      options: [
        {
          name: "Option name",
          modules: [
            {
              name: "Module name",
            },
          ],
        },
        {
          name: "Another option",
          modules: [
            {
              name: "Another module",
            },
          ],
        },
      ],
      description: "",
    },
  ],
};
