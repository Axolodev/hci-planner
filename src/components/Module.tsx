import { useCoursesStore } from "@/providers/CoursesStoreProvider";
import { Module as ModuleType } from "@/types";
import * as React from "react";

export const TESTING_ID = "Module";

interface Props extends ModuleType {
  moduleIndex: number;
  optionName: string;
  sectionTitle: string;
  optionIndex: number;
  disabled: boolean;
}

const Module: React.FunctionComponent<Props> = ({
  name,
  sectionTitle,
  optionIndex,
  moduleIndex,
  optionName,
  disabled,
}) => {
  const { getModuleStatus, setModuleStatus } = useCoursesStore(
    (state) => state
  );
  const isCompleted = getModuleStatus(sectionTitle, optionIndex, moduleIndex);
  const handleChange: React.ChangeEventHandler<HTMLInputElement> = (event) => {
    setModuleStatus(
      sectionTitle,
      optionIndex,
      moduleIndex,
      event.target.checked
    );
  };

  return (
    <input
      type="checkbox"
      onChange={handleChange}
      checked={isCompleted}
      disabled={disabled}
      id={`${optionName}-${name}`}
      className={`btn join-item disabled:cursor-not-allowed w-full rounded-field min-h-[40px] h-auto`}
      aria-label={name}
      style={{ margin: 0 }}
      data-testid={TESTING_ID}
    />
  );
};

export default Module;
