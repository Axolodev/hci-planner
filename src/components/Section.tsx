import { Fragment } from "react";
import ConfettiExplosion from "react-confetti-blast";

import { Section as SectionType } from "@/types";
import SectionOption from "./SectionOption";
import { useCoursesStore } from "@/providers/CoursesStoreProvider";

export const INDICATOR_TESTING_ID = "section-indicator";

interface Props {
  section: SectionType;
}

export default function Section({ section }: Props) {
  const { getIsSectionCompleted: isSectionCompleted } = useCoursesStore(
    (state) => state
  );
  const isCompleted = isSectionCompleted(section.title);

  return (
    <div className="p-4 bg-base-200 rounded-box flex flex-col indicator w-full">
      <label
        className={`swap text-3xl indicator-item absolute cursor-default ${
          isCompleted ? "swap-active" : ""
        }`}
        data-testid={INDICATOR_TESTING_ID}
      >
        <div className="swap-on">âœ…</div>
        <div className="swap-off"></div>
        {isCompleted && (
          <ConfettiExplosion
            force={0.3}
            duration={2200}
            particleCount={30}
            width={400}
            zIndex={2}
          />
        )}
      </label>

      <h2 className="text-xl font-bold pb-1">{section.title}</h2>
      <p className="pb-1">{section.description}</p>
      <div className="flex w-full gap-2 mt-2 flex-col md:flex-row flex-1">
        {section.options.map((option, index) => (
          <Fragment key={option.name}>
            {index > 0 && (
              <div className="divider md:divider-horizontal">OR</div>
            )}
            <SectionOption
              option={option}
              sectionTitle={section.title}
              optionIndex={index}
            />
          </Fragment>
        ))}
      </div>
    </div>
  );
}
