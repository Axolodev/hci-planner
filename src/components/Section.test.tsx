import React from "react";
import { render } from "@testing-library/react";
import { CoursesStoreProvider } from "../providers/CoursesStoreProvider";
import { TESTING_ID as MODULE_TESTING_ID } from "./Module";
import Section, { INDICATOR_TESTING_ID } from "./Section";
import { basicPlan } from "../testUtils/plans";
import userEvent from "@testing-library/user-event";

describe("Section", () => {
  it.skip("renders correctly and matches snapshot", () => {
    const { container } = render(
      <CoursesStoreProvider plan={basicPlan}>
        <Section section={basicPlan.sections[0]} />
      </CoursesStoreProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it("can check modules within a section, and shows an indicator when section is finished", async () => {
    const { getByTestId } = render(
      <CoursesStoreProvider plan={basicPlan}>
        <Section section={basicPlan.sections[0]} />
      </CoursesStoreProvider>
    );
    const checkbox = getByTestId(MODULE_TESTING_ID) as HTMLInputElement;
    const indicator = getByTestId(INDICATOR_TESTING_ID);

    expect(checkbox.checked).toBe(false);
    await userEvent.click(checkbox);
    expect(checkbox.checked).toBe(true);

    expect(indicator.classList.contains("swap-active")).toBe(true);
  });
});
