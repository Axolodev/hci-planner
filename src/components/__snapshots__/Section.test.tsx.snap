// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section renders correctly and matches snapshot 1`] = `
<div>
  <div
    class="p-4 bg-base-200 rounded-box flex flex-col indicator w-full"
  >
    <label
      class="swap text-3xl indicator-item absolute cursor-default "
      data-testid="section-indicator"
    >
      <div
        class="swap-on"
      >
        âœ…
      </div>
      <div
        class="swap-off"
      />
    </label>
    <h2
      class="text-xl font-bold pb-1"
    >
      Section title
    </h2>
    <p
      class="pb-1"
    />
    <div
      class="flex w-full gap-2 mt-2 flex-col md:flex-row flex-1"
    >
      <div
        class="p-4 flex-1 rounded-box flex flex-col items-stretch gap-2 justify-center transition-colors join w-full bg-base-300"
        data-testid="SectionOption-0"
      >
        <input
          aria-label="Module name"
          class="btn join-item disabled:cursor-not-allowed w-full rounded-field"
          data-testid="Module"
          id="Option name-Module name"
          style="margin: 0px;"
          type="checkbox"
        />
      </div>
    </div>
  </div>
</div>
`;
